[threaded-animations] turning on the "Threaded Scroll-driven Animations" setting should not force monotonic animations to be threaded

graouts opened the pull request:
#### f47e20b766b4c93246329c7b743b11e17cdf9b70
<pre>
[threaded-animations] turning on the &quot;Threaded Scroll-driven Animations&quot; setting should not force monotonic animations to be threaded
<a href="https://bugs.webkit.org/show_bug.cgi?id=302812">https://bugs.webkit.org/show_bug.cgi?id=302812</a>
<a href="https://rdar.apple.com/165066892">rdar://165066892</a>

Reviewed by Simon Fraser.

We introduced two different settings to control whether to run accelerated animations in the remote
layer tree: &quot;Threaded Scroll-driven Animations&quot; and &quot;Threaded Time-based Animations&quot;. However, we
don&apos;t actually support having one on without the other and thus turning on just &quot;Threaded Scroll-driven
Animations&quot; forces even time-based (ie. monotonic) animations to be accelerated via the remote layer
tree.

In this patch we correctly handle both settings individually. This starts in `KeyframeEffect` where
we introduce a new `canHaveAcceleratedRepresentation()` method which returns true if we can indeed
accelerate this animation via the remote layer tree while honoring the two aforementioned settings
individually. This essentially replaces the previous `threadedAnimationsEnabled()` method which
was the core reason we were in the situation described above.

We also use this new `canHaveAcceleratedRepresentation()` method when considering the creation
of an accelerated representation in `RenderLayerBacking::updateAcceleratedEffectsAndBaseValues()`,
ensuring that both the threaded and non-threaded animation flavors may be supported for a given
animation stack. Note that this does not account for the two flavors trying to animate the same
property, which will be handled in a followup patch by turning off acceleration altogether in
this situation.

To ensure that we only ever create accelerated timeline representations for monotonic and progress-based
timelines, we also add debug assertions in `DocumentTimeline` and `ScrollTimeline` to check the
relevant setting is enabled.

On the remote layer tree side, we must also change `RemoteLayerTreePropertyApplier::applyPropertiesToLayer()
in order to allow both animation flavors to be specified on a given remote layer tree node.

Finally, we add a new `isThreaded` member to the `AcceleratedAnimation` dictionary returned by the
`window.internals.acceleratedAnimationsForElement()` method so that we can determine the type of
accelerated animations created for a given element. We take this chance to update this testing-specific
method to be implemented within `AnimationTimelinesController` rather than `DocumentTimeline`
since this method has little to do with the document timeline.

Test: webanimations/threaded-animations/threaded-and-non-threaded-animations.html

* LayoutTests/webanimations/accelerated-animation-suspension.html:
* LayoutTests/webanimations/threaded-animations/threaded-and-non-threaded-animations-expected.txt: Added.
* LayoutTests/webanimations/threaded-animations/threaded-and-non-threaded-animations.html: Added.
* Source/WebCore/animation/AnimationTimelinesController.cpp:
(WebCore::AnimationTimelinesController::acceleratedAnimationsForElement const):
* Source/WebCore/animation/AnimationTimelinesController.h:
* Source/WebCore/animation/DocumentTimeline.cpp:
(WebCore::DocumentTimeline::createAcceleratedRepresentation):
(WebCore::DocumentTimeline::acceleratedAnimationsForElement const): Deleted.
* Source/WebCore/animation/DocumentTimeline.h:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::transformFunctionListPrefix const):
(WebCore::KeyframeEffect::isRunningAccelerated const):
(WebCore::KeyframeEffect::canBeAccelerated const):
(WebCore::KeyframeEffect::preventsAcceleration const):
(WebCore::KeyframeEffect::updateAcceleratedActions):
(WebCore::KeyframeEffect::addPendingAcceleratedAction):
(WebCore::KeyframeEffect::animationBecameReady):
(WebCore::KeyframeEffect::updateAcceleratedAnimationIfNecessary):
(WebCore::KeyframeEffect::animationDidFinish):
(WebCore::KeyframeEffect::animationWasCanceled):
(WebCore::KeyframeEffect::willChangeRenderer):
(WebCore::KeyframeEffect::animationSuspensionStateDidChange):
(WebCore::KeyframeEffect::applyPendingAcceleratedActionsOrUpdateTimingProperties):
(WebCore::KeyframeEffect::applyPendingAcceleratedActions):
(WebCore::KeyframeEffect::ticksContinuouslyWhileActive const):
(WebCore::KeyframeEffect::setComposite):
(WebCore::KeyframeEffect::effectStackNoLongerAllowsAccelerationDuringAcceleratedActionApplication):
(WebCore::KeyframeEffect::abilityToBeAcceleratedDidChange):
(WebCore::KeyframeEffect::lastStyleChangeEventStyleDidChange):
(WebCore::KeyframeEffect::canHaveAcceleratedRepresentation const):
(WebCore::KeyframeEffect::scheduleAssociatedAcceleratedEffectStackUpdate):
(WebCore::KeyframeEffect::threadedAnimationsEnabled const): Deleted.
* Source/WebCore/animation/KeyframeEffect.h:
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::createAcceleratedRepresentation):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::acceleratedAnimationsForTesting const):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::ensureStructuralLayer):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/texmap/coordinated/GraphicsLayerCoordinated.cpp:
(WebCore::GraphicsLayerCoordinated::acceleratedAnimationsForTesting const):
* Source/WebCore/platform/graphics/texmap/coordinated/GraphicsLayerCoordinated.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateAcceleratedEffectsAndBaseValues):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::acceleratedAnimationsForElement):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):

Canonical link: <a href="https://commits.webkit.org/303312@main">https://commits.webkit.org/303312@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f9c0bec64400199f63f52c735c39da05054753cc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132014 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/159/builds/4506 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43030 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/139528 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/83920 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/133884 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/155/builds/4449 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/156/builds/4268 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/139528 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/83920 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134960 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/155/builds/4449 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/118235 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/139528 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/155/builds/4449 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/macOS-Tahoe-Debug-API-Tests-EWS "Waiting to run tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/82748 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/155/builds/4449 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/36354 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/142175 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/153/builds/4176 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/36936 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/142175 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/160/builds/4257 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/156/builds/4268 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/142175 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/114512 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/57397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20525 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/158/builds/4229 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/32906 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/4061 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/67676 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/4321 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/4189 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/GraphicsLayer.h:
```diff
@@ -545,7 +545,7 @@ class GraphicsLayer : public RefCountedAndCanMakeWeakPtr<GraphicsLayer> {
     WEBCORE_EXPORT virtual void resumeAnimations();
 
-    virtual Vector<std::pair<String, double>> acceleratedAnimationsForTesting(const Settings&) const { return { }; }
+    virtual Vector<std::tuple<String, double, bool>> acceleratedAnimationsForTesting() const { return { }; }
```
Maybe it's time to make a struct for this return value.

graouts replied:
I think you spotted my laziness! I will do that in a followup.

graouts replied:
This is being done in https://github.com/WebKit/WebKit/pull/54238.

graouts committed:
[threaded-animations] turning on the "Threaded Scroll-driven Animations" setting should not force monotonic animations to be threaded
https://bugs.webkit.org/show_bug.cgi?id=302812
rdar://165066892

Reviewed by Simon Fraser.

We introduced two different settings to control whether to run accelerated animations in the remote
layer tree: "Threaded Scroll-driven Animations" and "Threaded Time-based Animations". However, we
don't actually support having one on without the other and thus turning on just "Threaded Scroll-driven
Animations" forces even time-based (ie. monotonic) animations to be accelerated via the remote layer
tree.

In this patch we correctly handle both settings individually. This starts in `KeyframeEffect` where
we introduce a new `canHaveAcceleratedRepresentation()` method which returns true if we can indeed
accelerate this animation via the remote layer tree while honoring the two aforementioned settings
individually. This essentially replaces the previous `threadedAnimationsEnabled()` method which
was the core reason we were in the situation described above.

We also use this new `canHaveAcceleratedRepresentation()` method when considering the creation
of an accelerated representation in `RenderLayerBacking::updateAcceleratedEffectsAndBaseValues()`,
ensuring that both the threaded and non-threaded animation flavors may be supported for a given
animation stack. Note that this does not account for the two flavors trying to animate the same
property, which will be handled in a followup patch by turning off acceleration altogether in
this situation.

To ensure that we only ever create accelerated timeline representations for monotonic and progress-based
timelines, we also add debug assertions in `DocumentTimeline` and `ScrollTimeline` to check the
relevant setting is enabled.

On the remote layer tree side, we must also change `RemoteLayerTreePropertyApplier::applyPropertiesToLayer()
in order to allow both animation flavors to be specified on a given remote layer tree node.

Finally, we add a new `isThreaded` member to the `AcceleratedAnimation` dictionary returned by the
`window.internals.acceleratedAnimationsForElement()` method so that we can determine the type of
accelerated animations created for a given element. We take this chance to update this testing-specific
method to be implemented within `AnimationTimelinesController` rather than `DocumentTimeline`
since this method has little to do with the document timeline.

Test: webanimations/threaded-animations/threaded-and-non-threaded-animations.html

* LayoutTests/webanimations/accelerated-animation-suspension.html:
* LayoutTests/webanimations/threaded-animations/threaded-and-non-threaded-animations-expected.txt: Added.
* LayoutTests/webanimations/threaded-animations/threaded-and-non-threaded-animations.html: Added.
* Source/WebCore/animation/AnimationTimelinesController.cpp:
(WebCore::AnimationTimelinesController::acceleratedAnimationsForElement const):
* Source/WebCore/animation/AnimationTimelinesController.h:
* Source/WebCore/animation/DocumentTimeline.cpp:
(WebCore::DocumentTimeline::createAcceleratedRepresentation):
(WebCore::DocumentTimeline::acceleratedAnimationsForElement const): Deleted.
* Source/WebCore/animation/DocumentTimeline.h:
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::transformFunctionListPrefix const):
(WebCore::KeyframeEffect::isRunningAccelerated const):
(WebCore::KeyframeEffect::canBeAccelerated const):
(WebCore::KeyframeEffect::preventsAcceleration const):
(WebCore::KeyframeEffect::updateAcceleratedActions):
(WebCore::KeyframeEffect::addPendingAcceleratedAction):
(WebCore::KeyframeEffect::animationBecameReady):
(WebCore::KeyframeEffect::updateAcceleratedAnimationIfNecessary):
(WebCore::KeyframeEffect::animationDidFinish):
(WebCore::KeyframeEffect::animationWasCanceled):
(WebCore::KeyframeEffect::willChangeRenderer):
(WebCore::KeyframeEffect::animationSuspensionStateDidChange):
(WebCore::KeyframeEffect::applyPendingAcceleratedActionsOrUpdateTimingProperties):
(WebCore::KeyframeEffect::applyPendingAcceleratedActions):
(WebCore::KeyframeEffect::ticksContinuouslyWhileActive const):
(WebCore::KeyframeEffect::setComposite):
(WebCore::KeyframeEffect::effectStackNoLongerAllowsAccelerationDuringAcceleratedActionApplication):
(WebCore::KeyframeEffect::abilityToBeAcceleratedDidChange):
(WebCore::KeyframeEffect::lastStyleChangeEventStyleDidChange):
(WebCore::KeyframeEffect::canHaveAcceleratedRepresentation const):
(WebCore::KeyframeEffect::scheduleAssociatedAcceleratedEffectStackUpdate):
(WebCore::KeyframeEffect::threadedAnimationsEnabled const): Deleted.
* Source/WebCore/animation/KeyframeEffect.h:
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::createAcceleratedRepresentation):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::acceleratedAnimationsForTesting const):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::ensureStructuralLayer):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/texmap/coordinated/GraphicsLayerCoordinated.cpp:
(WebCore::GraphicsLayerCoordinated::acceleratedAnimationsForTesting const):
* Source/WebCore/platform/graphics/texmap/coordinated/GraphicsLayerCoordinated.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateAcceleratedEffectsAndBaseValues):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::acceleratedAnimationsForElement):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):

Canonical link: https://commits.webkit.org/303312@main
